File: frontend/apps/ios/Pantopus/Features/Mailbox/MailboxDrawersViewModel.swift
//
//  MailboxDrawersViewModel.swift
//  Pantopus
//
//  Backs `MailboxDrawersView`. Fetches `GET /api/mailbox/v2/drawers` and
//  renders each drawer as a `file_chevron` row.
//

import Foundation
import Observation

/// ViewModel for the V2 mailbox drawers list.
@Observable
@MainActor
final class MailboxDrawersViewModel: ListOfRowsDataSource {
    let title = "Mailbox"
    /// A11.4 — surface the Mailbox map (physical postal venues) from the
    /// drawers root. The dedicated mailbox-root archetype (Wave B) will
    /// host this entry too once it lands.
    var topBarAction: TopBarAction? {
        TopBarAction(
            icon: .map,
            accessibilityLabel: "Map view"
        ) { [weak self] in
            Task { @MainActor in self?.onOpenMap() }
        }
    }

    let tabs: [ListOfRowsTab] = []
    var selectedTab: String = ""
    var fab: FABAction? {
        nil
    }

    private(set) var state: ListOfRowsState = .loading

    private let api: APIClient
    private let onOpenDrawer: (String) -> Void
    private let onOpenVault: () -> Void
    private let onOpenMap: @MainActor () -> Void

    init(
        api: APIClient = .shared,
        onOpenDrawer: @escaping (String) -> Void = { _ in },
        onOpenVault: @escaping () -> Void = {},
        onOpenMap: @escaping @MainActor () -> Void = {}
    ) {
        self.api = api
        self.onOpenDrawer = onOpenDrawer
        self.onOpenVault = onOpenVault
        self.onOpenMap = onOpenMap
    }

    func load() async {
        if case .loaded = state { return }
        state = .loading
        await fetch()
    }

    func refresh() async {
        await fetch()
    }

    func loadMoreIfNeeded() async {} // drawers are a fixed set.

    private func fetch() async {
        do {
            let response: DrawerListResponse = try await api.request(MailboxV2Endpoints.drawers())
            if response.drawers.isEmpty {
                state = .empty(
                    ListOfRowsState.EmptyContent(
                        icon: .mailbox,
                        headline: "No drawers yet",
                        subcopy: "Your drawers will show up here once mail arrives."
                    )
                )
            } else {
                // T6.5e (P19.5) — Mailbox root surfaces a "Vault" entry
                // alongside the four drawers so the saved-mail list is
                // reachable without leaving the inbox shell.
                let drawerRows = response.drawers.map(row(for:))
                let vaultRow = RowModel(
                    id: "vault",
                    title: "Vault",
                    subtitle: "Saved mail",
                    template: .fileChevron,
                    leading: .icon(.archive, tint: Theme.Color.primary600),
                    trailing: .chevron
                ) { @Sendable in Task { @MainActor in self.onOpenVault() } }
                state = .loaded(
                    sections: [RowSection(rows: drawerRows + [vaultRow])],
                    hasMore: false
                )
            }
        } catch {
            state = .error(message: (error as? APIError)?.errorDescription ?? "Couldn't load drawers.")
        }
    }

    private func row(for drawer: DrawerListResponse.Drawer) -> RowModel {
        let subtitle: String? = switch (drawer.unreadCount, drawer.urgentCount) {
        case (0, 0): nil
        case let (unread, 0): "\(unread) unread"
        case let (0, urgent): "\(urgent) urgent"
        case let (unread, urgent): "\(unread) unread · \(urgent) urgent"
        }

        let icon: PantopusIcon = switch drawer.drawer {
        case "personal": .user
        case "home": .home
        case "business": .shoppingBag
        case "earn": .megaphone
        default: .inbox
        }

        return RowModel(
            id: drawer.id,
            title: drawer.displayName,
            subtitle: subtitle,
            template: .fileChevron,
            leading: .icon(icon, tint: Theme.Color.primary600),
            trailing: .chevron
        ) { @Sendable in Task { @MainActor in self.onOpenDrawer(drawer.drawer) } }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/MailboxDrawersViewModel.kt
package app.pantopus.android.ui.screens.mailbox

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.mailbox.v2.Drawer
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.mailbox.MailboxRepository
import app.pantopus.android.ui.screens.shared.list_of_rows.ListOfRowsUiState
import app.pantopus.android.ui.screens.shared.list_of_rows.RowLeading
import app.pantopus.android.ui.screens.shared.list_of_rows.RowModel
import app.pantopus.android.ui.screens.shared.list_of_rows.RowSection
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTemplate
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTrailing
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for the V2 mailbox drawer list —
 * wraps `GET /api/mailbox/v2/drawers`.
 */
@HiltViewModel
class MailboxDrawersViewModel
    @Inject
    constructor(
        private val repo: MailboxRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<ListOfRowsUiState>(ListOfRowsUiState.Loading)

        /** Observed UI state. */
        val state: StateFlow<ListOfRowsUiState> = _state.asStateFlow()

        private var onOpenDrawer: (String) -> Unit = {}
        private var onOpenVault: () -> Unit = {}

        /** Wire nav callback before first load. */
        fun configureNavigation(
            onOpenDrawer: (String) -> Unit,
            onOpenVault: () -> Unit = {},
        ) {
            this.onOpenDrawer = onOpenDrawer
            this.onOpenVault = onOpenVault
        }

        /** Initial load. */
        fun load() {
            if (_state.value is ListOfRowsUiState.Loaded) return
            refresh()
        }

        /** Pull-to-refresh / retry. */
        fun refresh() {
            _state.value = ListOfRowsUiState.Loading
            viewModelScope.launch {
                when (val result = repo.drawers()) {
                    is NetworkResult.Success -> applySuccess(result.data.drawers)
                    is NetworkResult.Failure -> _state.value = ListOfRowsUiState.Error(result.error.message)
                }
            }
        }

        private fun applySuccess(drawers: List<Drawer>) {
            if (drawers.isEmpty()) {
                _state.value =
                    ListOfRowsUiState.Empty(
                        icon = PantopusIcon.Mailbox,
                        headline = "No drawers yet",
                        subcopy = "Your drawers will show up here once mail arrives.",
                    )
                return
            }
            // T6.5e (P19.5) — Append a "Vault" row alongside the four
            // drawers so the saved-mail list is reachable without
            // leaving the inbox shell.
            val drawerRows = drawers.map(::rowFor)
            val vaultRow =
                RowModel(
                    id = "vault",
                    title = "Vault",
                    subtitle = "Saved mail",
                    template = RowTemplate.FileChevron,
                    leading = RowLeading.Icon(icon = PantopusIcon.Archive, tint = PantopusColors.primary600),
                    trailing = RowTrailing.Chevron,
                    onTap = { onOpenVault() },
                )
            _state.value =
                ListOfRowsUiState.Loaded(
                    sections = listOf(RowSection(id = "drawers", rows = drawerRows + vaultRow)),
                    hasMore = false,
                )
        }

        private fun rowFor(drawer: Drawer): RowModel {
            val subtitle =
                when {
                    drawer.unreadCount == 0 && drawer.urgentCount == 0 -> null
                    drawer.urgentCount == 0 -> "${drawer.unreadCount} unread"
                    drawer.unreadCount == 0 -> "${drawer.urgentCount} urgent"
                    else -> "${drawer.unreadCount} unread · ${drawer.urgentCount} urgent"
                }
            val icon =
                when (drawer.drawer) {
                    "personal" -> PantopusIcon.User
                    "home" -> PantopusIcon.Home
                    "business" -> PantopusIcon.ShoppingBag
                    "earn" -> PantopusIcon.Megaphone
                    else -> PantopusIcon.Inbox
                }
            return RowModel(
                id = drawer.drawer,
                title = drawer.displayName,
                subtitle = subtitle,
                template = RowTemplate.FileChevron,
                leading = RowLeading.Icon(icon = icon, tint = PantopusColors.primary600),
                trailing = RowTrailing.Chevron,
                onTap = { onOpenDrawer(drawer.drawer) },
            )
        }
    }
